### PR TITLE
Allow client to specify an adapter to build an auth ctx from request

### PR DESCRIPTION
**Pull request by bwtimm**
Add a property on the Swatch-KOA API signature. Client can pass a custom function which will parse the request ctx and generate an 'auth ctx' object with custom auth information.

Also, updates the request handling pipeline to run any middleware functions as specified by the individual method before actually running the handler.  Pairs with https://github.com/builtforme/swatchjs/pull/4

### CODE REVIEW

File: lib/defaults.js [hunk 1]
```diff
@@ -29,4 +30,16 @@ function defaultPrefix(prefix) {
   return (prefix && `/${prefix}/`) || '/';
 }
 
+function defaultAuthAdapter(authAdapter) {
+  if (authAdapter && !(typeof authAdapter === 'function')) {
```
Comment (Furchin) [inline lib/defaults.js]: So I went down a rabbit hole last time I looked at this PR and lost track and here we are at the end of the week. 

So let's do it again!

I started off questioning why aren't you just using underscore for this -- and naturally underscore has [annotated source](http://underscorejs.org/docs/underscore.html#section-141) which shows they implement isFunction as:

```
_.each(['Arguments', 'Function', 'String', 'Number', 'Date', 'RegExp', 'Error'], function(name) {
    _['is' + name] = function(obj) {
      return toString.call(obj) === '[object ' + name + ']';
    };
  });
```

So obviously `typeof` and `toString.call` are different approaches to this. What's faster? A Google search comparing the two yields [this Stack Overflow answer](https://stackoverflow.com/questions/5999998/how-can-i-check-if-a-javascript-variable-is-function-type) which has a buried link to a [very interesting benchmark comparing four different approaches](http://jsben.ch/e5Ogr). While the links are a bit old, the benchmark runs in browser so it's interesting, yet re-running it several times the only consistent result is that jQuery is slowest by far. (In fact I've gotten each of the other three approaches to win at least once, although the constructor approach seems to rarely win.) 

For fun I added a fourth, to compare the `toString` approach underscore uses. It barely, but consistently, seems to beat jQuery for last place!

tl;dr: Good job :)
